larsim/Utils/TruthMatchUtils.h

```diff
@@ -0,0 +1,80 @@
+/**
+ * @file larsim/Utils/TruthMatchUtils.h
+ *
+ * @brief Utilities for matching a recob::Hit or vector of recob::Hit to the ID of the most significantly contributing G4 particle.  Various functions using different definitions of 'most significantly contributing' are provided.
+ *
+ * @author Dom Brailsford (d.brailsford@lancaster.ac.uk)
+ *
+ * $log: $
+ */
+#ifndef TRUTHMATCHUTILS_H_SEEN
+#define TRUTHMATCHUTILS_H_SEEN
+
+// LArSoft
+#include "lardataobj/RecoBase/Hit.h"
+
+// c++
+#include <map>
+#include <limits>
+#include <vector>
+
+namespace TruthMatchUtils {
+
+typedef int G4ID;
+typedef float EDeposit;
+typedef std::map<G4ID, EDeposit> IDToEDepositMap;
+
+const G4ID kNoG4ID = std::numeric_limits<G4ID>::lowest();           ///< The value used when no G4 ID has been found
```

Comment:
Generally not a good idea to put global variables in header files due to multiple definitions of the same variable.  If you have to do this, use an inline `constexpr` variable:
```.cpp
inline constexpr auto kNoG4ID = std::numeric_limits<G4ID>::lowest();
```

Reply:
There was a reason for doing this.  
The functions can return `kNoG4ID` when no G4 ID is found.  I figured it would be handy for end users to check if the function returned a sensible G4ID by comparing to `kNoG4ID`.  I also figured that as its wrapped in the namespace, the chance of a clash is minimal (provided users don't do something like `using namespace TruthMatchUtils` which I probably shouldn't assume...).

Is `constexpr` still fine in this case or should I do something like returning a literal from a function e.g.
```
inline constexpr G4ID TruthMatchUtils::NoG4ID() { return std::numeric_limits<G4ID>::lowest(); };
```

Reply:
I suspected you had a reason for using the `kNoG4ID` constant.  If, however, the purpose of it is to test validity, it would be better to create an actual function in the header file that people can call (e.g.):

```.cpp
namespace TruthMatchUtils {
  bool valid(G4ID id) noexcept;
  ...
}
```

and then in the implementation file (.cc):

```.cpp
namespace {
  // Implementation detail, local to file
  constexpr auto invalid_id = std::numeric_limits<G4ID>::lowest();
}

namespace TruthMatchUtils {
  bool valid(G4ID const id) noexcept {
    return invalid_id != id;
  }
  ...
}
```

This way, users of the header simply do something like:
```.cpp
auto const id = TrueParticleIDFromTotalRecoCharge(...);
if (valid(id)) {
  // do something
}
```
which is clearer than an explicit comparison like:
```.cpp
if (id != kG4NoID) {
  // do something
}
```

Reply:
Totally agree, good idea.  I've implemented this.